### PR TITLE
COMPRESS-477 building a split zip

### PR DESCRIPTION
**Pull request by PeterAlfreadLee**
[COMPRESS-477](https://issues.apache.org/jira/projects/COMPRESS/issues/COMPRESS-477)
Add support for building a split/spanned zip.

Sample code:
```
@Test
public void buildSplitZipTest() throws IOException {
	File directoryToZip = getFilesToZip();
	File outputZipFile = new File(dir, "splitZip.zip");
	long splitSize = 100 * 1024L; /* 100 KB */
	final ZipArchiveOutputStream zipArchiveOutputStream = new ZipArchiveOutputStream(outputZipFile, splitSize);

	addFilesToZip(zipArchiveOutputStream, directoryToZip);
	zipArchiveOutputStream.close();
	// TODO: validate the created zip files when extracting split zip is merged into master
}

private void addFilesToZip(ZipArchiveOutputStream zipArchiveOutputStream, File fileToAdd) throws IOException {
	if(fileToAdd.isDirectory()) {
		for(File file : fileToAdd.listFiles()) {
			addFilesToZip(zipArchiveOutputStream, file);
		}
	} else {
		ZipArchiveEntry zipArchiveEntry = new ZipArchiveEntry(fileToAdd.getPath());
		zipArchiveEntry.setMethod(ZipEntry.DEFLATED);

		zipArchiveOutputStream.putArchiveEntry(zipArchiveEntry);
		IOUtils.copy(new FileInputStream(fileToAdd), zipArchiveOutputStream);
		zipArchiveOutputStream.closeArchiveEntry();
	}
}
```

This PR is implemented by adding a new class `ZipSplitOutputStream`, and it's mainly implemented like this:
1. Write the zip split signature to the zip file in the constructor of `ZipSplitOutputStream` by calling `writeZipSplitSignature`;
2. Based on the zip specification, the split size must between 64K and 4,294,967,295 bytes;
3. Rename the split zip files like .z01, .z02, ... , .z(N-1), .zip ONLY IF there are more than 1 split segment;
4. Get the only split segment whose suffix is .zip IF the split size is big enough(it means the split size is bigger than the actual zip size);
5. Create a new zip split segment if the size of data to write exceeds split size, and the newly created zip segment will be named in the sequence like .z01, .z02, ..., .z99, .z100, .z101, ... , .zip;
6. Based on the zip specification, the End Of Central Directory(EOCD) and Zip64 End Of Central Directory Locator(Zip64_EOCDL) must reside on the same segment, so the `ZipSplitOutputStream` will create a new segment if the remaining size is not enough before writing EOCD and Zip64_EOCDL;
7. When creating `ZipArchiveOutputStream`, if the split size is specified, it will create a split zip instead of normal zip(as the `ZipSplitOutputStream` need the file name when creating new split segments, the constructor is like `public ZipArchiveOutputStream(final File file, final long zipSplitSize)`);
8. The disk number, relative offset, number of this disk, number of Central Directories on this disk, total number of disks in Central Directory, Zip64 End Of Central Directory, Zip64 End Of Central Directory Locator, End Of Central Directory have all been tuned to the right value when writing a split/spanned zip;
9. The testcases need to be updated when [#84]{https://github.com/apache/commons-compress/pull/84} is merged because it seems I can not test my created split/spanned zip in Linux. I have tested it on Windows and it works well;
10. This PR has some minor conflicts with [#84]{https://github.com/apache/commons-compress/pull/84}, and I will solve all these conflicts as soon as [#84]{https://github.com/apache/commons-compress/pull/84} is merged.

Please feel free to let me know if the code need to be refactored or rebased. I'm looking for your reviews. :-) @bodewig @garydgregory

### CODE REVIEW

**Comment by PeterAlfreadLee** [comment]
The travis-ci failed when building on openjdk12, and all the others are passed.
I tested my code on Ubuntu with oracle jdk8 by `mvn verify` and it's passed.
I checked the log and I think it's a problem caused by network, but I'm not sure about this.
Is this a problem caused by my code?

**Comment by bodewig** [comment]
@PeterAlfreadLee this looks very good overall, many thanks. I've added a few review comments. Apart from that some `@since` markers and javadocs would be nice.

I'm not sure whether the split feature and parallel creation of archives could be combined, I don't think it would be possible but need to ponder it a bit longer. If it is indeed not possible, we should document the fact.

**Comment by PeterAlfreadLee** [comment]
@bodewig Thank you so much.
Unfortunely, just like the other PR, I can't see any comments in `File changed` panel of thie PR. I'm sorry if I missed something. Would you please tell me where I can see the comments if I missed the comments?

> I'm not sure whether the split feature and parallel creation of archives could be combined, I don't think it would be possible but need to ponder it a bit longer. If it is indeed not possible, we should document the fact.

I'll look into this.

**Comment by bodewig** [comment]
For me there is a "@bodewig started a review" section right on this conversation page and I see my comments inline inside of the Files panel. But if I look at things in a separate browser without being signed in I don't see anything either. Strange.

I'll try to figure things out.

**Comment by bodewig** [comment]
Sorry, I misunderstood github's code review feature and wasn't aware I was the only one who could see it until I "sumitted" the review. There was no indication that the review is only visible to myself, sorry about that.

**Comment by PeterAlfreadLee** [comment]
I can see the comments now. I will fix all these ASAP.

File: src/main/java/org/apache/commons/compress/compressors/FileNameUtil.java [hunk 1]
```diff
@@ -193,4 +193,19 @@ public String getCompressedFilename(final String fileName) {
         return fileName + defaultExtension;
     }
 
+    public static String getBaseName(String filename) {
+        if (filename == null) {
+            return null;
+        }
+
+        int lastSeparatorPosition = Math.max(filename.lastIndexOf('/'), filename.lastIndexOf('\\'));
+        String name = filename.substring(lastSeparatorPosition + 1);
```
[the remark this answers is not in the document]
Comment (PeterAlfreadLee) [reply src/main/java/org/apache/commons/compress/compressors/FileNameUtil.java]: Agree

File: src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStream.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        String baseName = FileNameUtil.getBaseName(zipFile.getName());
+        String extension = ".z";
+        if (newZipSplitSegmentSuffixIndex <= 9) {
+            extension += "0" + newZipSplitSegmentSuffixIndex;
+        } else {
+            extension += newZipSplitSegmentSuffixIndex;
+        }
+
+        String newFileName = zipFile.getParent() + File.separatorChar + baseName + extension;
+        File newFile = new File(newFileName);
```
[the remark this answers is not in the document]
Comment (PeterAlfreadLee) [reply src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStream.java]: Agree

File: src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStream.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+import org.apache.commons.compress.compressors.FileNameUtil;
+
+import java.io.*;
+import java.nio.ByteBuffer;
+
+public class ZipSplitOutputStream extends OutputStream {
+    private OutputStream outputStream;
+    private File zipFile;
+    private long splitSize;
```
[the remark this answers is not in the document]
Comment (PeterAlfreadLee) [reply src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStream.java]: Agree

File: src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStream.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+        long bytesRemainingInThisSegment = this.splitSize - this.currentSplitSegmentBytesWritten;
+        if (bytesRemainingInThisSegment < unsplittableContentSize) {
+            openNewSplitSegment();
+        }
+    }
+
+    @Override
+    public void write(int i) throws IOException {
+        byte[] b = ByteBuffer.allocate(4).putInt(i).array();
```
[the remark this answers is not in the document]
Comment (PeterAlfreadLee) [reply src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStream.java]: Agree. This should no doubt be a single byte. Sorry for my careless.

[the remark this answers is not in the document]
Comment (bodewig) [reply src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStream.java]: no worries, that's what a second pair of eyes is good for.

**Comment by PeterAlfreadLee** [comment]
Code just pushed with all these reviews fixed.

**Comment by bodewig** [comment]
https://github.com/bodewig/commons-compress/tree/COMPRESS-477-constructing holds a rebased version of this PR, I'd be happy if you just applied similar changes and pushed it here, no need to have me as contributor inside of your work.

While rebasing I realized that master says the disk number start is a long while this branch uses an int. Actually both master and this branch neglect the possibility of the number of disks requiring Zip64 extra handling. TBH I find the idea of a split ZIP archive that spans more than 64k files a bit disturbing and would address it with a separate change independent of this PR.

**Comment by trejkaz** [comment]
I'll be keeping watch on this one as we're very interested in the functionality. How soon can we get this in an official release?

**Comment by PeterAlfreadLee** [comment]
@bodewig 
Thank you so much for your reviews. I'll check this rebased version out ASAP.
I'll at least add your name to the comments to thank you for your hard work.

**Comment by bodewig** [comment]
@trejkaz I won't make any promises but I intend to propose a new release once this PR and #87 are in - and I've looked through the latest JIRA issues. Most likely I won't be able to start with this before christmas, but hope to cut 1.20 early next year.

**Comment by PeterAlfreadLee** [comment]
> Actually both master and this branch neglect the possibility of the number of disks requiring Zip64 extra handling. TBH I find the idea of a split ZIP archive that spans more than 64k files a bit disturbing and would address it with a separate change independent of this PR.

You are right. I forgot to take the Zip64 into consideration. I'll also look into this later. And I agree that a separate PR maybe a good idea.

**Comment by PeterAlfreadLee** [comment]
Code pushed. Just the same as https://github.com/bodewig/commons-compress/tree/COMPRESS-477-constructing with only a little change : the imports of ZipSplitOutputStream has been extracted.

Besides I'll add some more test cases. Unfortunely I got some other work this week so I don't have that much time for OSS like last few weeks. I hope I can finish it in these two days. Thank you for your reviews @bodewig

**Comment by PeterAlfreadLee** [comment]
Code pushed with some test cases added.

**Comment by bodewig** [comment]
@PeterAlfreadLee the openjdk-ea build is failing in master as well. It probably either is a problem that exists in master anyway or a problem with our Travis setup (which happens more often than our code breaking anything).

**Comment by bodewig** [comment]
@PeterAlfreadLee please take a look at my changes in 26c7aff 

I think all that needs to be done now is document split support inside of the "manual" properly and then I can close COMPRESS-477. I'll take care of that.

**Comment by PeterAlfreadLee** [comment]
@bodewig 
Thank you for your reviews and rebased code.

> I think all that needs to be done now is document split support inside of the "manual" properly and then I can close COMPRESS-477. I'll take care of that.

Please feel free to let me know if I can offer some help.

**Comment by bodewig** [comment]
@PeterAlfreadLee please have a look at 57664d91 and check I haven't made any mistakes.

**Comment by PeterAlfreadLee** [comment]
Everything looks great. I will try to finish the part of Zip64 for split zip ASAP.

**Pull request merged**